---
title: "BIOINFO-615"
author: "Hernan Lorenzi"
date: "3/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Farhat please set the PATH below to your working directory
knitr::opts_knit$set(root.dir = "/Users/lorenziha/OneDrive/PROJECTS/BIOINFO-615/FARHAT_DATA/STAR_DATA/")
```

---
title: "RNAseq training - Farhat"
author: Hernan Lorenzi
output: html_document
---

Farhat, to run the R code you first have to set the path to the working directory where you will have the test datasets. That path have to be stored in the variable "my_path" on line 27 below

## Load libraries and define functions

```{r load libraries and define functions}

library("org.Hs.eg.db")
library("pheatmap")
library("EnhancedVolcano")
library("ggplot2")
library("ggpubr")
library("DESeq2")
library("stringr")

# Define function to generate plots
plot_gene <- function(gene_name){
  
  # retrieve gene_id from gene_name
  keep <- human_gene_annotation$Gene.name == gene_name
  gene_id     <- rownames(human_gene_annotation[keep, ])
  
  d <- plotCounts(dds, gene = gene_id, intgroup="Diff", 
                  returnData=TRUE)
  
  ggplot(d, aes(x=Diff, y=count)) + 
    geom_point(position=position_jitter(w=0.1,h=0)) + 
    scale_y_log10(breaks=c(25,100,400)) + ggtitle(paste(gene_id, gene_name))
  
  ggboxplot(d, x = "Diff", 
              y = "count",
              color = "Diff", 
              palette =c("#00AFBB", "#E7B800", "#FC4E07", "#CC00FF", "#32DCFF"),
              add = "jitter", 
              shape = "Diff",
              yscale = "log10",
              title = paste(gene_id , gene_name)
            )
  
  ggviolin(d, x = "Diff", 
           y = "count", 
           fill = "Diff",
           palette = c("#00AFBB", "#E7B800", "#FC4E07", "#CC00FF", "#32DCFF"),
           add = "boxplot", 
           add.params = list(fill = "white"),
           yscale = "log10",
           title = paste(gene_id, gene_name)
           )
}


replace_gene_acc_by_symbols_in_dds <- function(res.tmp){
  ensemble_ids <- rownames(res.tmp)
  symbols <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL') #, multiVals = "first"
  symbols <- symbols[!is.na(symbols)]
  to_name <- rownames(res.tmp) %in% names(symbols)
  res.tmp@rownames[to_name] <- as.vector(symbols)
  return(res.tmp)
}

# Replace ensembl_ids by genbank gene_ids.
replace_gene_acc_by_gb_ids <- function(ensemble_ids, return_all = TRUE){
  entrezids <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('ENTREZID'), keytype = 'ENSEMBL', multiVals = "first")
  entrezids <- entrezids[!is.na(entrezids)]
  to_name <- ensemble_ids %in% names(entrezids)
  ensemble_ids[to_name] <- as.vector(entrezids)
  if (return_all){
    return(ensemble_ids)
  }
  else {
    return(ensemble_ids[to_name])
  }
}

replace_gene_acc_by_symbol_ids <- function(ensemble_ids, return_all = TRUE){
  entrezids <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL', multiVals = "first")
  entrezids <- entrezids[!is.na(entrezids)]
  to_name <- ensemble_ids %in% names(entrezids)
  ensemble_ids[to_name] <- as.vector(entrezids)
  if (return_all){
    return(ensemble_ids)
  }
  else {
    return(ensemble_ids[to_name])
  }
}


replace_ensemble_acc_by_symbols <- function(ensemble_ids){
  symbols <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL', multiVals = "first")
  symbols <- symbols[!is.na(symbols)]
  to_name <- ensemble_ids %in% names(symbols)
  ensemble_ids[to_name] <- as.vector(symbols)
  return(ensemble_ids)
}

generate_volcano_plot <- function(res.tmp, my_file_name, log_scale = FALSE){
  res.tmp <- replace_gene_acc_by_symbols_in_dds(res.tmp)
  vp <- EnhancedVolcano(res.tmp, 
                        lab = rownames(res.tmp), 
                        x = 'log2FoldChange', 
                        y = 'pvalue',
                        pointSize = 1,
                        colAlpha = 4/5,
                        labSize = 2,  # Controls labels size
                        title = res.tmp@elementMetadata$description[2],
                        titleLabSize = 14,
                        subtitle = '', # add subtitle here
                        subtitleLabSize = 12,
                        legendPosition = 'right',
                        legendLabSize = 12,
                        legendIconSize = 2.0,
                        axisLabSize = 10
                        )
  
  if (log_scale){
    vp <- vp + scale_x_log10()
  }
  ggsave(filename = paste0(my_file_name,".pdf"), plot = vp )
  
  print(vp)
}

generate_volcano_plot_with_ids <- function(res.tmp, my_file_name, log_scale = FALSE, gene_list){
  res.tmp <- replace_gene_acc_by_symbols_in_dds(res.tmp)
  vp <- EnhancedVolcano(res.tmp, 
                        lab = rownames(res.tmp), 
                        x = 'log2FoldChange', 
                        y = 'pvalue',
                        pointSize = 1,
                        colAlpha = 4/5,
                        labSize = 4,  # Controls labels size
                        labCol = "black",
                        labFace = 'bold',
                        title = res.tmp@elementMetadata$description[2],
                        titleLabSize = 14,
                        subtitle = '', # add subtitle here
                        subtitleLabSize = 12,
                        legendPosition = 'right',
                        legendLabSize = 14,
                        legendIconSize = 4.0,
                        axisLabSize = 10,
                        drawConnectors = TRUE,
                        widthConnectors = 0.75,
                        colConnectors = "black",
                        selectLab = gene_list, # vector of gene symbols to label on volcanoplot
                        boxedLabels = FALSE
                        )
  
  if (log_scale){
    vp <- vp + scale_x_log10()
  }
  ggsave(filename = paste0(my_file_name,".pdf"), plot = vp )
  
  print(vp)
}

plot_heat_map <- function( my_vstd, gene_list, file_name, variables){
  
  # Replace gene IDs by gene symbols
  gene_list <- replace_ensemble_acc_by_symbols(gene_list)
  rownames(my_vstd) <- replace_ensemble_acc_by_symbols(rownames(my_vstd))
  
  # Plot the heat map
  hmp <- pheatmap(assay(my_vstd)[gene_list,], cluster_rows=T, show_rownames=TRUE,
           cluster_cols=T, annotation_col = variables, fontsize_col = 5, tile = file_name)
  
  ggsave(filename = paste0(file_name,"_heatmap.pdf"), plot = hmp, width = 8.5, height = 11, units = "in")
  
  print(hmp)
}

### It will take list of gene symbols as gene_list
plot_heat_map_from_gene_symbols <- function( my_vstd, gene_list, file_name, variables){
  
  # Replace gene IDs by gene symbols
  #gene_list <- replace_ensemble_acc_by_symbols(gene_list)
  rownames(my_vstd) <- replace_ensemble_acc_by_symbols(rownames(my_vstd))
  
  # Plot the heat map
  hmp <- pheatmap(assay(my_vstd)[gene_list,], cluster_rows=T, show_rownames=TRUE,
           cluster_cols=T, annotation_col = variables, fontsize_col = 5, tile = file_name)
  
  ggsave(filename = paste0(file_name,"_heatmap.pdf"), plot = hmp, width = 8.5, height = 11, units = "in")
  
  print(hmp)
}

```

## Load data

```{r Loading data}

all <- read.delim2("all_counts.txt", sep = "\t", header = TRUE, row.names = 1, comment.char = c("#"))

# Remove ".sorted.dedup.bam" suffix from colnames and .# suffix from ensembl IDs.
colnames(all) <- str_remove(string = colnames(all),pattern = ".sorted.dedup.bam")
rownames(all) <- str_remove(string = rownames(all), pattern = '.[0-9]+$')
all <- all[,7:ncol(all)] # Delete cols with no read counts

# Keep table with Ensemble IDs and gene Symbols
ensembl_to_symbol <- as.data.frame(cbind("Ensembl_ID" = rownames(all), "Symbol" = all$gene_name), row.names = 1)

# Load metadata
metadata <- read.delim2("all.counts_metadata.txt", sep = "\t", row.names = 1, header = T)

## Sort tables so metadata and read counts match order
metadata <- metadata[order(rownames(metadata)),]
all <- all[,order(colnames(all))]


#groups = metadata$experiment # see file goups.txt

#Remove all zero rows
all=all[rowSums(all)>0,]

```


###########################################
## Analysis of expression data using DESeq2
###########################################

```{r Diff expression function}

# Add column with gene symbols

diff_exp_function <- function( dds_object, my_file_name, column_of_interest, case, control, alpha_cutoff){
  # Extract differential expression results based on "host type" using "NS" as numerator and "M" as denominator for Log2FC calculation.
  my_results <- results(dds_object, contrast=c(column_of_interest,case,control), alpha = alpha_cutoff) # Compare ND vs M
  
  # Sort results by adj.p-value
  my_results <- my_results[order(my_results$padj),]
  
  #print(my_results)
  
  # Get coefficient names
  #resultsNames(dds_object)
  
  # Adjust Log2FC to control for genes with low counts and high dispersion
  my_coef = paste0(column_of_interest, "_", case, "_vs_", control)
  my_results <- lfcShrink(dds_object, coef = my_coef, type="apeglm")
  
  # Sort results by FDR (padj)
  my_results <- my_results[order(my_results$padj),]
  #print(my_results)
  
  # sumarize results
  #summary(my_results, alpha = 0.05)
  
  # MA-plots of results
  #print(plotMA(my_results, ylim=c(-10, 10)))
  
  diff_expression_table.df <- as.data.frame(my_results)
  
  # Add column with gene symbols and with entrez gene IDs
  diff_expression_table.df$gene_symbols <- replace_gene_acc_by_symbol_ids(ensemble_ids = rownames(diff_expression_table.df), return_all = TRUE)
  
  diff_expression_table.df$gene_entrez_ids <- replace_gene_acc_by_gb_ids(ensemble_ids = rownames(diff_expression_table.df), return_all = TRUE)
  # Save differential expression results to a file
  write.table(diff_expression_table.df, file = my_file_name, sep = "\t")
  
  return(my_results)
}
```



```{r prepare data}

# Convert conditions to be compared to factors
metadata$Diff = as.factor(as.character(metadata$Diff))
metadata$Donor_id = as.factor(as.character(metadata$Donor_id))
metadata$State = as.factor(as.character(metadata$State))

# Generate DESeq2 object for ST and NS conditions together
#dds.paired <- DESeqDataSetFromMatrix(countData = all, 
#                              colData = metadata,  
#                              design = ~ Donor_id + State + Diff) # controling by strain type


# Subset all and metadata tables so they contain either NS or ST samples 
keep <- metadata$State == "NS"
all.NS <- all[ , keep ] # read counts for NS
all.ST <- all[ , !keep ] # read counts for ST
metadata.NS <- metadata[ keep , ] # metadata for NS
metadata.ST <- metadata[ !keep , ] # Metadata for ST

# Generate DESeq2 object for NS and ST condition ONLY
dds.ns <- DESeqDataSetFromMatrix(countData = all.NS, 
                              colData = metadata.NS,  
                              design = ~ Donor_id + Diff)

dds.st <- DESeqDataSetFromMatrix(countData = all.ST, 
                              colData = metadata.ST,  
                              design = ~ Donor_id + Diff)


## Prefiltering genes with less than 10 counts in at least 2 samples
keep <- rowSums(counts(dds.ns) > 10) >= 2
dds.ns <- dds.ns[keep,]

keep <- rowSums(counts(dds.ns) > 10) >= 2
dds.ns <- dds.ns[keep,]


# Set CCR6_Neg as reference for all comparisons
dds.ns$Diff <- relevel(dds.ns$Diff, "CCR6_Neg")
dds.st$Diff <- relevel(dds.st$Diff, "CCR6_Neg")

# Run DESeq2 on dds object    
dds.ns <- DESeq(dds.ns)
dds.st <- DESeq(dds.st)

# Transform read counts for PCA analysis and plotting heatmaps
vstd.ns <- vst(dds.ns)
vstd.st <- vst(dds.st)

# Generate result tables for NS and ST conditions.

# p-value cutoff
my_alpha <- 0.99

res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_Hi_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_Hi", control = "CCR6_Neg", alpha_cutoff = my_alpha)

res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_Low_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_Low", control = "CCR6_Neg", alpha_cutoff = my_alpha)

res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_PosCCR2_Pos", control = "CCR6_Neg", alpha_cutoff = my_alpha)

res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_Hi_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_Hi", control = "CCR6_Neg", alpha_cutoff = my_alpha)

res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_Low_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_Low", control = "CCR6_Neg", alpha_cutoff = my_alpha)

res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_PosCCR2_Pos", control = "CCR6_Neg", alpha_cutoff = my_alpha)

```



```{r plot heat maps}


# Select top 20 most differentially expressed genes between CCR6_Hi and CCR6_Neg 
genes_of_interest.ns <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns)[1:20]
genes_of_interest.st <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.st)[1:20]

# For all genes in your dataset
# genes_of_interest.ns <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns)

# Select factors to be displayed together with the expression data
df.ns <- as.data.frame(colData(dds.ns)[,c("Diff","State")])
df.st <- as.data.frame(colData(dds.st)[,c("Diff","State")])

# plot_heat_map( list_of_genes_to_plot , file_name_for_heatmap, 1 or 2 variables to annotate in the heatmap)
plot_heat_map(my_vstd = vstd.ns, gene_list = genes_of_interest.ns, file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns", variables = df.ns)
plot_heat_map(my_vstd = vstd.st, gene_list = genes_of_interest.st, file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns", variables = df.st)

# The function below will take a list of gene symbols for generating the heatmaps.
plot_heat_map_from_gene_symbols(my_vstd = vstd.ns, gene_list = c("MAP3K4", "FURIN","AGPAT4","THEM5","C2CD4A","CCL20","SELL","THBS1", "CCR6","CCR2"), file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns_TEST", variables = df.ns)

```
********  Heat Map:Replace sample IDs by sample_names values

```{r volcano plot}


# generate_volcano_plot( result , file_name_for_plot )
generate_volcano_plot(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns, "res_CCR6_Hi_vs_CCR6_Neg_NS_volcano_plot")

#######################################################
# Add parameter to specify what genes to highlight
#######################################################

# generate_volcano_plot( result , file_name_for_plot, list of genes to label )
generate_volcano_plot_with_ids(res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.ns, "res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_volcano_plot", gene_list = c("CCR2", "CCR6"))


# Farhat, I added a parameter to the generate_volcano_plot function, log_scale, to use a log10 scale on the X axis if set to TRUE. I realize that the issue with that is the negative LogFC cannot be represented but you will see that the possitive values get scattered around.  
generate_volcano_plot(res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.ns, "res_CCR6_Low_vs_CCR6_Neg_NS_volcano_plot", log_scale = FALSE)

```

#################################
## Gene / GO enrichment analysis
#################################

The code below select a list of gene IDs based on a specific Log2FC cutoff for running Overrepresentation analysis with the R function 'enricher'. The list of genes are stored in the variable "gene" for gene IDs = Ensembl gene IDs, and in the variable  'gene.entrezid' for a list of Entrez gene IDs. Usually, geneset databases use eithr Entrez gene IDs or gene symbols as identifiers, so in some cases it is handy to count with a list of genes using Entrez gene IDs. 

The same code also extract the entire list of genes used for the Differential Expression Analysis as an R list element, using Ensemble gene IDs as keys (names) and Log2FC as values. This geneList is used as input for running Gene Set Enrichment Analysis with the function 'GSEA'. If needed, the code below also generates the same geneList but using Entrez gene IDs as keys.

```{r Select genes for enrichment analysis}
library("clusterProfiler")

# Change the line below to run the enrichment analysis on any datasets, for example: 
# geneList <- res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.st[,2]
# For stimulated condition.
geneList <- res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.ns[,2]
geneList <- sort(geneList, decreasing = TRUE) # Sort geneList by Log2FC

# Convert Ensembl IDs to EntrezIDs
geneList.entrezid <- geneList
names(geneList.entrezid) <- replace_gene_acc_by_gb_ids(names(geneList), return_all = TRUE)

# Discard genes missing Entrez Gene IDs
keep <- grep('ENSG', names(geneList.entrezid), invert = TRUE)
geneList.entrezid <- geneList.entrezid[keep]

# Set cutoff for Log2FC below (ideally you should have between 500 and 3000 genes for the enricher analysis)

# For entrez gene IDs
log2FC_cutoff <- 0.4
gene.entrezid <- names(geneList.entrezid)[abs(geneList.entrezid) > log2FC_cutoff]
print(paste( "Total number of Entrez gene IDs = ",length(gene.entrezid)))

# For ensemble Genen IDs
gene <- names(geneList)[abs(geneList) > log2FC_cutoff]
print(paste( "Total number of Ensemble gene IDs = ", length(gene)))


```

## Enrichment analysis with MSigDB

Farhat, you can see what MSigDBs are available from download here: https://www.gsea-msigdb.org/gsea/msigdb/

MSigDBs can be downloaded from R with the function msigdbr (see below). You can learn more about this function by typing in the R Console '?msigdbr()' or by adding the function name in the Help window. However, for some useful DBs I could not figure out the name of the subcategory, and hence, I couldn't download them from R. In those cases, the next best approach is to manually download the DB from the website to your computer and then upload the gene set with the read.table (or any other equivalent) function in R and then use it in your gene enrichment analysis. 

```{r MSigDB}
library(msigdbr)
library(cowplot)

# Plotting functions
draw_enrichment_barplot <- function(my_enricher_obj, 
                                    my_pathway_counts = 10, 
                                    file_name_prefix = "enricher_barplot", 
                                    my_width = 11, my_height = 8){
  my_enricher_obj@result$p.adjust <- as.numeric(format(my_enricher_obj@result$p.adjust,  digits=3))
  sp <- ggbarplot(head(my_enricher_obj@result, n = my_pathway_counts), 
            x = "Description", 
            y = "Count",
            fill = "p.adjust",          # change fill color by cyl
            color = "white",            # Set bar border colors to white
            sort.val = "desc",          # Sort the value in dscending order
            sort.by.groups = FALSE,     # Don't sort inside each group
            x.text.angle = 90,          # Rotate vertically x axis texts
            rotate = TRUE,
            ggtheme = theme_minimal(),
            ylab = c("Gene counts"),
            ) + gradient_fill("RdYlBu")
  print(sp)
  ggsave2(filename = paste0(file_name_prefix,".pdf"), plot = sp, width = my_width, height = my_height)
}


draw_GSEA_barplot <- function(my_gsea_obj, 
                              my_pathway_counts = 10, 
                              file_name_prefix = "gsea_barplot", 
                              my_width = 11, my_height = 8){
  my_gsea_obj@result$p.adjust <- as.numeric(format(my_gsea_obj@result$p.adjust,  digits=3))
  sp <- ggbarplot(head(my_gsea_obj@result, n = my_pathway_counts), 
            x = "Description", 
            y = "NES",
            fill = "p.adjust",          # change fill color by cyl
            color = "white",            # Set bar border colors to white
            sort.val = "desc",          # Sort the value in dscending order
            sort.by.groups = FALSE,     # Don't sort inside each group
            x.text.angle = 90,          # Rotate vertically x axis texts
            rotate = TRUE,
            ggtheme = theme_minimal(),
            ylab = c("Normalized Enrichment Score (NES)"),
            ) + gradient_fill("RdYlBu")
  print(sp)
  ggsave2(filename = paste0(file_name_prefix,".pdf"), plot = sp, width = my_width, height = my_height)
}

# You can modify the category and subcategory parameters below to retrieve specific gene sets from GSEA's MSIG database.

########################################
## C7 immunologic signature gene sets
########################################
msig_c7 <- msigdbr(species = "Homo sapiens", category = "C7") %>% dplyr::select(gs_name, ensembl_gene, ) 

# TO Run overrepresentation analysis
msig_c7.oa <- enricher(gene, TERM2GENE=msig_c7,  pvalueCutoff=0.99)

# TO Run Gene Set Enrichment Analysis
msig_c7.gsea <- GSEA(geneList, TERM2GENE = msig_c7, pvalueCutoff = 0.99, verbose = TRUE)

head(msig_c7.oa, n=1000)
head(msig_c7.gsea, n=1000)

# Convert Entrez gene IDs to gene symbols
msig_c7.oa.gs <- setReadable(msig_c7.oa, org.Hs.eg.db, keyType = "ENSEMBL")
head(msig_c7.oa.gs, n=1000)
msig_c7.gsea.gs <- setReadable(msig_c7.gsea, org.Hs.eg.db, keyType = "ENSEMBL")
head(msig_c7.gsea.gs, n=1000)

# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(msig_c7.oa.gs), file = "OR_msigdb.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(msig_c7.gsea.gs), file = "GSEA_msigdb.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

# Generate plots summarizing enrichment results (You might need to tweak the height and width parameters of the plot to make it fit within the page limits and look nice, otherwise it might shrink to fit within the page)
library(enrichplot)

# Dotplot
p1 <- dotplot(msig_c7.oa.gs, showCategory=30) + ggtitle("dotplot for ORA")
p2 <- dotplot(msig_c7.gsea.gs, showCategory=30) + ggtitle("dotplot for GSEA")
print(p1)
print(p2)
ggsave2(filename = "OA_msig_c7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_dotplot.pdf", plot = p1, width = 11, height = 8)
ggsave2(filename = "GSEA_msig_c7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_dotplot.pdf", plot = p2, width = 11, height = 8)

# Barplot
draw_enrichment_barplot(my_enricher_obj = msig_c7.oa.gs, my_pathway_counts = 15, file_name_prefix = "OA_msig_c7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_barplot", my_width = 11, my_height = 8)

draw_GSEA_barplot(my_gsea_obj =  msig_c7.gsea.gs, my_pathway_counts = 15, file_name_prefix = "GSEA_msig_c7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_barplot", my_width = 11, my_height = 8)

#######################################################
## C2 curated gene sets => CP: Canonical pathways
#######################################################

msig_c2_cp <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP") %>% dplyr::select(gs_name, ensembl_gene, ) 

em <- enricher(gene, TERM2GENE = msig_c2_cp,  pvalueCutoff=0.99)
em2 <- GSEA(geneList, TERM2GENE = msig_c2_cp, pvalueCutoff = 0.99, verbose = TRUE)

head(em, n=1000)
head(em2, n=1000)

# Convert Entrez gene IDs to gene symbols
em.gs <- setReadable(em, org.Hs.eg.db, keyType = "ENSEMBL")
head(em.gs, n=1000)
em2.gs <- setReadable(em2, org.Hs.eg.db, keyType = "ENSEMBL")
head(em2.gs, n=1000)

# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(em.gs), file = "OR_msigdb.C2_CP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(em2.gs), file = "GSEA_msigdb.C2_CP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

#######################################################
## C3 regulatory target gene sets => TFT: transcription factor targets
#######################################################

msig_c3_gtrd <- msigdbr(species = "Homo sapiens", category = "C3", subcategory = "TFT:GTRD") %>% dplyr::select(gs_name, ensembl_gene, )
msig_c3_tft <- msigdbr(species = "Homo sapiens", category = "C3") %>% dplyr::select(gs_name, ensembl_gene, )
# Remove MIR gene sets
keep <- grep('MIR[0-9]',msig_c3_tft$gs_name, invert = T)
msig_c3_tft <- msig_c3_tft[keep,]

c3.gtrd.or <- enricher(gene, TERM2GENE=msig_c3_gtrd,  pvalueCutoff=0.99)
c3.gtrd.gsea <- GSEA(geneList, TERM2GENE = msig_c3_gtrd, pvalueCutoff = 0.99, verbose = TRUE)

c3.tft.or <- enricher(gene, TERM2GENE=msig_c3_tft,  pvalueCutoff=0.99)
c3.tft.gsea <- GSEA(geneList, TERM2GENE = msig_c3_tft, pvalueCutoff = 0.99, verbose = TRUE)

# Convert Entrez gene IDs to gene symbols
c3.gtrd.or <- setReadable(c3.gtrd.or, org.Hs.eg.db, keyType = "ENSEMBL")
head(c3.gtrd.or, n=1000)
c3.gtrd.gsea <- setReadable(c3.gtrd.gsea, org.Hs.eg.db, keyType = "ENSEMBL")
head(c3.gtrd.gsea, n=1000)

c3.tft.or <- setReadable(c3.tft.or, org.Hs.eg.db, keyType = "ENSEMBL")
head(c3.tft.or, n=1000)
c3.tft.gsea <- setReadable(c3.tft.gsea, org.Hs.eg.db, keyType = "ENSEMBL")
head(c3.tft.gsea, n=1000)

# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(c3.gtrd.or), file = "OR_msigdb.C3_GTRD_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(c3.gtrd.gsea), file = "GSEA_msigdb.C3_GTRD_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

write.table(as.data.frame(c3.tft.or), file = "OR_msigdb.C3_TFT_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(c3.tft.gsea), file = "GSEA_msigdb.C3_TFT_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

```

WikiPathways analysis: 

```{r WikiPathways analysis}
library(magrittr)
library(tidyr)

# Extracting gene set info from database (WikiPathways)
wpgmtfile <- system.file("extdata/wikipathways-20180810-gmt-Homo_sapiens.gmt", package="clusterProfiler")
wp2gene <- data.frame(read.gmt(wpgmtfile))
wp2gene <- wp2gene %>% tidyr::separate(term, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME


############################################################
# Farhat, I would leave the p-val and qval cutoffs in the functions below as they are set right now to get as many pathways as possoble in the output and then you can filtered out with Excel.
############################################################

# Overrepresentation analysis
ewp <- enricher(gene.entrezid, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, pvalueCutoff=0.99, qvalueCutoff = 0.15)
head(ewp, n=1000)

# Convert Entrez gene IDs to gene symbols
ewp.gs <- setReadable(ewp, org.Hs.eg.db, keyType = "ENTREZID")
head(ewp.gs, n=1000)

# x <- res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns$log2FoldChange
# x <- sort(x, decreasing = T)
# ewp2 <- GSEA(x, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, verbose=FALSE)

# Gene enrichment analysis with GSEA
ewp2 <- GSEA(geneList.entrezid, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, verbose=FALSE, pvalueCutoff = 0.15)
head(ewp2)

# To generate plot
#library(enrichplot)
#gseaplot2(ewp2, geneSetID = "WP727", title = ewp2$Description[1])

# Convert Entrez gene IDs to gene symbols
ewp2.gs <- setReadable(ewp2, org.Hs.eg.db, keyType = "ENTREZID")
head(ewp2.gs, n=1000)


# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(ewp.gs), file = "OR_WikiPathways_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp2.gs), file = "GSEA_WikiPathways_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
```





```{r GO}
ggo <- groupGO(gene     = gene,
               OrgDb    = org.Hs.eg.db,
               ont      = "BP",  # Either MF, BP or CC
               level    = 3, # The higher the number, the more specific the GO terms.
               readable = TRUE, 
               keyType = 'ENSEMBL'
               )

head(ggo, n = 1000)

ego <- enrichGO(gene          = gene,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego), n = 1000)

ego.mf <- enrichGO(gene          = gene,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "MF", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )

# Looking at up or downregulated genes
gene_up <- names(geneList[geneList >= 0.5])
gene_down <- names(geneList[geneList <= -0.5])

ego.up <- enrichGO(gene          = gene_up,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego.up), n = 1000)


ego.down <- enrichGO(gene     = gene_down,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego.down), n = 1000)

# Write output to a file (GG = GroupGO, EGO = Enrichment GO. EGO.UP = enrichment GO for upregulated genes; EGO.DN = enrichment GO for downregulated genes). Change file names for diff conditions.
write.table(as.data.frame(ggo), file = "GG.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ego), file = "EGO_BP.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ego.up), file = "EGO.UP_BP.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ego.down), file = "EGO.DN_BP.C7_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

write.table(as.data.frame(ego.mf), file = "EGO_MF.CCR6_PosCCR2_Pos_vs_CCR6_Neg_ns.txt", sep = "\t")

# Barplot
# To control how many categories are included in the plot you have to tweak then number using the my_pathway_counts parameter.
draw_enrichment_barplot(my_enricher_obj = ego, my_pathway_counts = 15, file_name_prefix = "OA_GO_BP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_barplot", my_width = 11, my_height = 8)

# To control how many categories are included in the plot you have to tweak then number using the howCategory parameter.
p1 <- dotplot(ego, showCategory=30) + ggtitle("dotplot for GO BP enrichment")
print(p1)

ggsave2(filename = "OA_GO_BP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_dotplot.pdf", plot = p1, width = 11, height = 8)

# Upregulated genes 
draw_enrichment_barplot(my_enricher_obj = ego.up, my_pathway_counts = 15, file_name_prefix = "OA_GO_BP_UP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_barplot", my_width = 11, my_height = 8)

p1 <- dotplot(ego.up, showCategory=30) + ggtitle("dotplot for GO BP enrichment (upregulated genes)")
print(p1)

ggsave2(filename = "OA_GO_BP_UP_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_dotplot.pdf", plot = p1, width = 11, height = 8)

# Downregulated genes 

# MF
draw_enrichment_barplot(my_enricher_obj = ego.mf, my_pathway_counts = 15, file_name_prefix = "OA_GO_MF_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_barplot", my_width = 11, my_height = 8)

p1 <- dotplot(ego.mf, showCategory=30) + ggtitle("dotplot for GO MF enrichment")
print(p1)

ggsave2(filename = "OA_GO_MF_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS_dotplot.pdf", plot = p1, width = 11, height = 8)

```
  
  
  

