---
title: "BIOINFO-615"
author: "Hernan Lorenzi"
date: "3/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Farhat please set the PATH below to your working directory
knitr::opts_knit$set(root.dir = "/Users/lorenziha/OneDrive/PROJECTS/BIOINFO-615/FARHAT_DATA/")
```

---
title: "RNAseq analysis of experimental infections for Rima"
author: Hernan Lorenzi
output: html_document
---

Farhat, to run the R code you first have to set the path to the working directory where you will have the test datasets. That path have to be stored in the variable "my_path" on line 27 below

## Load libraries and define functions

```{r load libraries and define functions}

library("org.Hs.eg.db")
library("pheatmap")
library("EnhancedVolcano")
library("ggplot2")
library("ggpubr")
library("DESeq2")

# Define function to generate plots
plot_gene <- function(gene_name){
  
  # retrieve gene_id from gene_name
  keep <- human_gene_annotation$Gene.name == gene_name
  gene_id     <- rownames(human_gene_annotation[keep, ])
  
  d <- plotCounts(dds, gene = gene_id, intgroup="Diff", 
                  returnData=TRUE)
  
  ggplot(d, aes(x=Diff, y=count)) + 
    geom_point(position=position_jitter(w=0.1,h=0)) + 
    scale_y_log10(breaks=c(25,100,400)) + ggtitle(paste(gene_id, gene_name))
  
  ggboxplot(d, x = "Diff", 
              y = "count",
              color = "Diff", 
              palette =c("#00AFBB", "#E7B800", "#FC4E07", "#CC00FF", "#32DCFF"),
              add = "jitter", 
              shape = "Diff",
              yscale = "log10",
              title = paste(gene_id , gene_name)
            )
  
  ggviolin(d, x = "Diff", 
           y = "count", 
           fill = "Diff",
           palette = c("#00AFBB", "#E7B800", "#FC4E07", "#CC00FF", "#32DCFF"),
           add = "boxplot", 
           add.params = list(fill = "white"),
           yscale = "log10",
           title = paste(gene_id, gene_name)
           )
}


replace_gene_acc_by_symbols_in_dds <- function(res.tmp){
  ensemble_ids <- rownames(res.tmp)
  symbols <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL') #, multiVals = "first"
  symbols <- symbols[!is.na(symbols)]
  to_name <- rownames(res.tmp) %in% names(symbols)
  res.tmp@rownames[to_name] <- as.vector(symbols)
  return(res.tmp)
}

# Replace ensembl_ids by genbank gene_ids.
replace_gene_acc_by_gb_ids <- function(ensemble_ids, return_all = TRUE){
  entrezids <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('ENTREZID'), keytype = 'ENSEMBL', multiVals = "first")
  entrezids <- entrezids[!is.na(entrezids)]
  to_name <- ensemble_ids %in% names(entrezids)
  ensemble_ids[to_name] <- as.vector(entrezids)
  if (return_all){
    return(ensemble_ids)
  }
  else {
    return(ensemble_ids[to_name])
  }
}

replace_gene_acc_by_symbol_ids <- function(ensemble_ids, return_all = TRUE){
  entrezids <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL', multiVals = "first")
  entrezids <- entrezids[!is.na(entrezids)]
  to_name <- ensemble_ids %in% names(entrezids)
  ensemble_ids[to_name] <- as.vector(entrezids)
  if (return_all){
    return(ensemble_ids)
  }
  else {
    return(ensemble_ids[to_name])
  }
}


replace_ensemble_acc_by_symbols <- function(ensemble_ids){
  symbols <- mapIds(org.Hs.eg.db, keys = ensemble_ids, column = c('SYMBOL'), keytype = 'ENSEMBL', multiVals = "first")
  symbols <- symbols[!is.na(symbols)]
  to_name <- ensemble_ids %in% names(symbols)
  ensemble_ids[to_name] <- as.vector(symbols)
  return(ensemble_ids)
}

generate_volcano_plot <- function(res.tmp, my_file_name, log_scale = FALSE){
  res.tmp <- replace_gene_acc_by_symbols_in_dds(res.tmp)
  vp <- EnhancedVolcano(res.tmp, 
                        lab = rownames(res.tmp), 
                        x = 'log2FoldChange', 
                        y = 'pvalue',
                        pointSize = 1,
                        colAlpha = 4/5,
                        labSize = 2,  # Controls labels size
                        title = res.tmp@elementMetadata$description[2],
                        titleLabSize = 14,
                        subtitle = '', # add subtitle here
                        subtitleLabSize = 12,
                        legendPosition = 'right',
                        legendLabSize = 12,
                        legendIconSize = 2.0,
                        axisLabSize = 10
                        )
  
  if (log_scale){
    vp <- vp + scale_x_log10()
  }
  ggsave(filename = paste0(my_file_name,".pdf"), plot = vp )
  
  print(vp)
}

plot_heat_map <- function( my_vstd, gene_list, file_name, variables){
  
  # Replace gene IDs by gene symbols
  gene_list <- replace_ensemble_acc_by_symbols(gene_list)
  rownames(my_vstd) <- replace_ensemble_acc_by_symbols(rownames(my_vstd))
  
  # Plot the heat map
  hmp <- pheatmap(assay(my_vstd)[gene_list,], cluster_rows=T, show_rownames=TRUE,
           cluster_cols=T, annotation_col = variables, fontsize_col = 5, tile = file_name)
  
  ggsave(filename = paste0(file_name,"_heatmap.pdf"), plot = hmp, width = 8.5, height = 11, units = "in")
  
  print(hmp)
}

### It will take list of gene symbols as gene_list
plot_heat_map_from_gene_symbols <- function( my_vstd, gene_list, file_name, variables){
  
  # Replace gene IDs by gene symbols
  #gene_list <- replace_ensemble_acc_by_symbols(gene_list)
  rownames(my_vstd) <- replace_ensemble_acc_by_symbols(rownames(my_vstd))
  
  # Plot the heat map
  hmp <- pheatmap(assay(my_vstd)[gene_list,], cluster_rows=T, show_rownames=TRUE,
           cluster_cols=T, annotation_col = variables, fontsize_col = 5, tile = file_name)
  
  ggsave(filename = paste0(file_name,"_heatmap.pdf"), plot = hmp, width = 8.5, height = 11, units = "in")
  
  print(hmp)
}

```

## Loading data

```{r set working dir}
#my_path = "/Users/lorenziha/OneDrive/PROJECTS/BIOINFO-615/FARHAT_DATA/"
#setwd(my_path)
```


```{r Loading data}

all <- read.delim2("all.counts.txt", sep = "\t", header = TRUE, row.names = 1)

metadata <- read.delim2("all.counts_metadata.txt", sep = "\t", row.names = 1, header = T)

## Sort tables so metadata and read counts match order
metadata <- metadata[order(rownames(metadata)),]
all <- all[,order(colnames(all))]

# Remove summary rows from read counts table "all"
all <- all[grep("_",rownames(all), invert = T),]

groups = metadata$experiment # see file goups.txt

human_gene_annotation <- read.delim2("human_gene_annotation.txt", sep = "\t", header = TRUE, row.names = 1)

#Remove all zero rows
all=all[rowSums(all)>0,]

```

If you want to explore your expression data without having to deal with a lot of R code, you can use the following R package:

## pcaExplorer

This will open up an interactive window where you can generate PCA plots, heat maps and look for genes that are differentially perturbed accross conditions of interest. You can learn more about pcaExplorer in the documentation page here: https://federicomarini.github.io/pcaExplorer/articles/pcaExplorer.html

For running pcaExplorer with the test dataset you can do the following:

```{r}
library(pcaExplorer)
pcaExplorer(countmatrix = all_h, coldata = metadata)

```

###########################################
## Analysis of expression data using DESeq2
###########################################

```{r prepare data}

# Convert conditions to be compared to factors
metadata$Diff = as.factor(as.character(metadata$Diff))

# Prepare DESeqDataSet object (dds) including all required metadata and read counts per gene 
#read_count_table <- all #all_m_ns_ni
#metadata_table = metadata #meta_m_ns_ni

# Generate DESeq2 object
dds <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~Diff)



## Prefiltering genes with less than 10 counts in at least 2 samples
keep <- rowSums(counts(dds) > 10) >= 2
dds <- dds[keep,]

```

## Run DESeq2
res_ND_vs_M and res_NS_vs_M objects contain the Differential Gene Expression (DGE) results using an alpha cutoff of 0.1. You can customize the alpha value to whaterver works for your analysis.

The PCA plot below can also be customized to, for example, use different symbols to represent other factors of interest.
```{r run DESeq2}
# Set CCR6_Neg as reference for all comparisons
dds$Diff <- relevel(dds$Diff, "CCR6_Neg")

# Run DESeq2 on dds object    
dds <- DESeq(dds)

# Transform read counts for PCA analysis and plotting heatmaps
vstd <- vst(dds)

# draw PCA coloring by "host" type
plotPCA(vstd, intgroup = "Diff", ntop = 1000, returnData = FALSE)
plotPCA(vstd, intgroup = "Donor_id", ntop = 1000, returnData = FALSE)
plotPCA(vstd, intgroup = "State", ntop = 1000, returnData = FALSE)
plotPCA(vstd, intgroup = "Replicate", ntop = 1000, returnData = FALSE)

# Extract differential expression results based on "host type" using "NS" as numerator and "M" as denominator for Log2FC calculation.
alpha_cutoff = 0.1 
res_CCR6_Hi_vs_CCR6_Neg <- results(dds, contrast=c("Diff","CCR6_Hi","CCR6_Neg"), alpha = alpha_cutoff) # Compare ND vs M
res_CCR6_Low_vs_CCR6_Neg <- results(dds, contrast=c("Diff","CCR6_Low","CCR6_Neg"), alpha = alpha_cutoff) # Compare NS vs M

# Sort results by adj.p-value
res_CCR6_Hi_vs_CCR6_Neg <- res_CCR6_Hi_vs_CCR6_Neg[order(res_CCR6_Hi_vs_CCR6_Neg$padj),]
res_CCR6_Low_vs_CCR6_Neg <- res_CCR6_Low_vs_CCR6_Neg[order(res_CCR6_Low_vs_CCR6_Neg$padj),]

# print out results to the screen
print(res_CCR6_Hi_vs_CCR6_Neg)
print(res_CCR6_Low_vs_CCR6_Neg)


```


```{r manipulate results}
# Get coefficient names
resultsNames(dds)

# Adjust Log2FC to control for genes with low counts and high dispersion
res_CCR6_Hi_vs_CCR6_Neg_LFC <- lfcShrink(dds, coef="Diff_CCR6_Hi_vs_CCR6_Neg", type="apeglm")

# Sort results by FDR (padj)
res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted <- res_CCR6_Hi_vs_CCR6_Neg_LFC[order(res_CCR6_Hi_vs_CCR6_Neg_LFC$padj),]
print(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted)

# sumarize results
summary(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted, alpha = 0.05)

# MA-plots of results
plotMA(res_CCR6_Hi_vs_CCR6_Neg, ylim=c(-10, 10))
plotMA(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted, ylim=c(-10, 10))

# with 
plotMA(res_CCR6_Low_vs_CCR6_Neg, ylim=c(-10, 10))

```
```{r other plots}


# Call plot_gene function with gene name of interest, for instance:
# plot_gene(gene_name)

plot_gene("MT-RNR2")
plot_gene("RORC")

```

```{r exporting results to a file}
write.csv(as.data.frame(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted), file="res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.csv")
```

```{r plot heat maps}
library("pheatmap")
# Select top 20 most expressed genes across all samples
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
# Select factors to be displayed together with the expression data
df <- as.data.frame(colData(dds)[,c("Diff","State")])

# Plot the heat map
pheatmap(assay(vstd)[select,], cluster_rows=T, show_rownames=TRUE,
         cluster_cols=T, annotation_col=df, fontsize_col = 5)

```


```{r Diff expression function}

# Add column with gene symbols

diff_exp_function <- function( dds_object, my_file_name, column_of_interest, case, control, alpha_cutoff){
  # Extract differential expression results based on "host type" using "NS" as numerator and "M" as denominator for Log2FC calculation.
  my_results <- results(dds_object, contrast=c(column_of_interest,case,control), alpha = alpha_cutoff) # Compare ND vs M
  
  # Sort results by adj.p-value
  my_results <- my_results[order(my_results$padj),]
  
  #print(my_results)
  
  # Get coefficient names
  #resultsNames(dds_object)
  
  # Adjust Log2FC to control for genes with low counts and high dispersion
  my_coef = paste0(column_of_interest, "_", case, "_vs_", control)
  my_results <- lfcShrink(dds_object, coef = my_coef, type="apeglm")
  
  # Sort results by FDR (padj)
  my_results <- my_results[order(my_results$padj),]
  #print(my_results)
  
  # sumarize results
  #summary(my_results, alpha = 0.05)
  
  # MA-plots of results
  #print(plotMA(my_results, ylim=c(-10, 10)))
  
  diff_expression_table.df <- as.data.frame(my_results)
  
  # Add column with gene symbols and with entrez gene IDs
  diff_expression_table.df$gene_symbols <- replace_gene_acc_by_symbol_ids(ensemble_ids = rownames(diff_expression_table.df), return_all = TRUE)
  
  diff_expression_table.df$gene_entrez_ids <- replace_gene_acc_by_gb_ids(ensemble_ids = rownames(diff_expression_table.df), return_all = TRUE)
  # Save differential expression results to a file
  write.table(diff_expression_table.df, file = my_file_name, sep = "\t")
  
  return(my_results)
}
```




```{r paired analysis}

# Convert conditions to be compared to factors
metadata$Donor_id = as.factor(as.character(metadata$Donor_id))

# Generate DESeq2 object for ST and NS conditions together
dds.paired <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ Donor_id + State + Diff) # controling by strain type


# Subset all and metadata tables so they contain either NS or ST samples 
keep <- metadata$State == "NS"
all.NS <- all[ , keep ] # read counts for NS
all.ST <- all[ , !keep ] # read counts for ST
metadata.NS <- metadata[ keep , ] # metadata for NS
metadata.ST <- metadata[ !keep , ] # Metadata for ST

# Generate DESeq2 object for NS and ST condition ONLY
dds.ns <- DESeqDataSetFromMatrix(countData = all.NS, 
                              colData = metadata.NS,  
                              design = ~ Donor_id + Diff)

dds.st <- DESeqDataSetFromMatrix(countData = all.ST, 
                              colData = metadata.ST,  
                              design = ~ Donor_id + Diff)


## Prefiltering genes with less than 10 counts in at least 2 samples
keep <- rowSums(counts(dds.ns) > 10) >= 2
dds.ns <- dds.ns[keep,]

keep <- rowSums(counts(dds.ns) > 10) >= 2
dds.ns <- dds.ns[keep,]


# Set CCR6_Neg as reference for all comparisons
dds.ns$Diff <- relevel(dds.ns$Diff, "CCR6_Neg")
dds.st$Diff <- relevel(dds.st$Diff, "CCR6_Neg")

# Run DESeq2 on dds object    
dds.ns <- DESeq(dds.ns)
dds.st <- DESeq(dds.st)

# Transform read counts for PCA analysis and plotting heatmaps
vstd.ns <- vst(dds.ns)
vstd.st <- vst(dds.st)

# Generate result tables for NS and ST conditions.
res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_Hi_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_Hi", control = "CCR6_Neg", alpha_cutoff = 0.1)

res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_Low_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_Low", control = "CCR6_Neg", alpha_cutoff = 0.1)

res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.ns <- diff_exp_function(dds_object = dds.ns, my_file_name = "res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_NS.txt", column_of_interest = "Diff", case = "CCR6_PosCCR2_Pos", control = "CCR6_Neg", alpha_cutoff = 0.1)

res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_Hi_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_Hi", control = "CCR6_Neg", alpha_cutoff = 0.1)

res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_Low_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_Low", control = "CCR6_Neg", alpha_cutoff = 0.1)

res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_LFC_sorted.st <- diff_exp_function(dds_object = dds.st, my_file_name = "res_CCR6_PosCCR2_Pos_vs_CCR6_Neg_ST.txt", column_of_interest = "Diff", case = "CCR6_PosCCR2_Pos", control = "CCR6_Neg", alpha_cutoff = 0.1)

```



```{r plot heat maps}


# Select top 20 most differentially expressed genes between CCR6_Hi and CCR6_Neg 
genes_of_interest.ns <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns)[1:20]
genes_of_interest.st <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.st)[1:20]

# FOr all genes in your dataset
# genes_of_interest.ns <- rownames(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns)

# Select factors to be displayed together with the expression data
df.ns <- as.data.frame(colData(dds.ns)[,c("Diff","State")])
df.st <- as.data.frame(colData(dds.st)[,c("Diff","State")])

# plot_heat_map( list_of_genes_to_plot , file_name_for_heatmap, 1 or 2 variables to annotate in the heatmap)
plot_heat_map(my_vstd = vstd.ns, gene_list = genes_of_interest.ns, file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns", variables = df.ns)
plot_heat_map(my_vstd = vstd.st, gene_list = genes_of_interest.st, file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns", variables = df.st)

# The function below will take a list of gene symbols for generating the heatmaps.
plot_heat_map_from_gene_symbols(my_vstd = vstd.ns, gene_list = c("MAP3K4", "FURIN","AGPAT4","THEM5","C2CD4A","CCL20","SELL","THBS1"), file_name = "res_CCR6_Hi_vs_CCR6_Neg_ns_TEST", variables = df.ns)

```
********  Heat Map:Replace sample IDs by sample_names values

```{r volcano plot}


# generate_volcano_plot( result , file_name_for_plot )
generate_volcano_plot(res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns, "res_CCR6_Hi_vs_CCR6_Neg_NS_volcano_plot")

# Farhat, I added a parameter to the generate_volcano_plot function, log_scale, to use a log10 scale on the X axis if set to TRUE. I realize that the issue with that is the negative LogFC cannot be represented but you will see that the possitive values get scattered around.  
generate_volcano_plot(res_CCR6_Low_vs_CCR6_Neg_LFC_sorted.ns, "res_CCR6_Low_vs_CCR6_Neg_NS_volcano_plot", log_scale = FALSE)

```

#################################
## Gene / GO enrichment analysis
#################################

```{r Select genes for enrichment analysis}
library("clusterProfiler")

geneList <- res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns[,2]
geneList <- sort(geneList, decreasing = TRUE) # Sort geneList by Log2FC

# Convert Ensembl IDs to EntrezIDs
geneList.entrezid <- geneList
names(geneList.entrezid) <- replace_gene_acc_by_gb_ids(names(geneList), return_all = TRUE)

# Discard genes missing Entrez Gene IDs
keep <- grep('ENSG', names(geneList.entrezid), invert = TRUE)
geneList.entrezid <- geneList.entrezid[keep]

# Set cutoff for Log2FC
# For entrez gene IDs
gene.entrezid <- names(geneList.entrezid)[abs(geneList.entrezid) > 0.15]
print(paste( "Entrez gene IDs = ",length(gene.entrezid)))

# For ensemble Genen IDs
gene <- names(geneList)[abs(geneList) > 0.2]
print(paste( "Ensemble gene IDs = ", length(gene)))


```

WikiPathways analysis

```{r WikiPathways analysis}
library(magrittr)
library(tidyr)

# Extracting gene set info from database (WikiPathways)
wpgmtfile <- system.file("extdata/wikipathways-20180810-gmt-Homo_sapiens.gmt", package="clusterProfiler")
wp2gene <- data.frame(read.gmt(wpgmtfile))
wp2gene <- wp2gene %>% tidyr::separate(term, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME


############################################################
# Farhat, I would leave the p-val and qval cutoffs in the functions belowas they are set right now to get as many pathways as possoble ain the output and then you can filtered out with Excel.
############################################################

# Overrepresentation analysis
ewp <- enricher(gene.entrezid, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, pvalueCutoff=0.99, qvalueCutoff = 0.99)
head(ewp, n=1000)

# Convert Entrez gene IDs to gene symbols
ewp.gs <- setReadable(ewp, org.Hs.eg.db, keyType = "ENTREZID")
head(ewp.gs, n=1000)

# x <- res_CCR6_Hi_vs_CCR6_Neg_LFC_sorted.ns$log2FoldChange
# x <- sort(x, decreasing = T)
# ewp2 <- GSEA(x, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, verbose=FALSE)

# Gene enrichment analysis with GSEA
ewp2 <- GSEA(geneList.entrezid, TERM2GENE = wpid2gene, TERM2NAME = wpid2name, verbose=FALSE, pvalueCutoff = 0.99)
head(ewp2)

# To generate plot
#library(enrichplot)
#gseaplot2(ewp2, geneSetID = "WP727", title = ewp2$Description[1])

# Convert Entrez gene IDs to gene symbols
ewp2.gs <- setReadable(ewp2, org.Hs.eg.db, keyType = "ENTREZID")
head(ewp2.gs, n=1000)


# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(ewp.gs), file = "OR_WikiPathways_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp2.gs), file = "GSEA_WikiPathways_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
```



```{r MSigDB}
library(msigdbr)

# You can modify the category and subcategory parameters below to retrieve specific gene sets from GSEA's MSIG database.
msig_c2_cp_reactome <- msigdbr(species = "Homo sapiens", category = "C7") %>% dplyr::select(gs_name, ensembl_gene, ) 

em <- enricher(gene, TERM2GENE=msig_c2_cp_reactome,  pvalueCutoff=0.99)
em2 <- GSEA(geneList, TERM2GENE = msig_c2_cp_reactome, pvalueCutoff = 0.99, verbose = TRUE)

head(em, n=1000)
head(em2, n=1000)

# Convert Entrez gene IDs to gene symbols
em.gs <- setReadable(em, org.Hs.eg.db, keyType = "ENSEMBL")
head(em.gs, n=1000)

# Convert Entrez gene IDs to gene symbols
em2.gs <- setReadable(em2, org.Hs.eg.db, keyType = "ENSEMBL")
head(em2.gs, n=1000)

# Write output to a file (OR stands for Overrepresenation analysis). Change file names for diff conditions.
write.table(as.data.frame(ewp.gs), file = "OR_msigdb.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp2.gs), file = "GSEA_msigdb.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")

```

```{r GO}
ggo <- groupGO(gene     = gene,
               OrgDb    = org.Hs.eg.db,
               ont      = "BP",  # Either MF, BP or CC
               level    = 3, # The higher the number, the more specific the GO terms.
               readable = TRUE, 
               keyType = 'ENSEMBL'
               )

head(ggo, n = 1000)

ego <- enrichGO(gene          = gene,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.99,
                qvalueCutoff  = 0.99,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego), n = 1000)

# Looking at up or downregulated genes
gene_up <- names(geneList[geneList >= 0.5])
gene_down <- names(geneList[geneList <= -0.5])

ego.up <- enrichGO(gene          = gene_up,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.99,
                qvalueCutoff  = 0.99,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego.up), n = 1000)


ego.down <- enrichGO(gene     = gene_down,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP", # Either MF, BP or CC
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.99,
                qvalueCutoff  = 0.99,
                keyType = 'ENSEMBL',
                readable      = TRUE
                )
head(as.data.frame(ego.down), n = 1000)

# Write output to a file (GG = GroupGO, EGO = Enrichment GO. EGO.UP = enrichment GO for upregulated genes; EGO.DN = enrichment GO for downregulated genes). Change file names for diff conditions.
write.table(as.data.frame(ewp.gs), file = "GG.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp2.gs), file = "EGO_BP.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp.gs), file = "EGO.UP_BP.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")
write.table(as.data.frame(ewp2.gs), file = "EGO.DN_BP.C7_CCR6_Hi_vs_CCR6_Neg_ns.txt", sep = "\t")

```


